agent/config/builder.go

```diff
@@ -1580,6 +1586,7 @@ func (b *builder) checkVal(v *CheckDefinition) *structs.CheckDefinition {
 		TTL:                            b.durationVal(fmt.Sprintf("check[%s].ttl", id), v.TTL),
 		SuccessBeforePassing:           intVal(v.SuccessBeforePassing),
 		FailuresBeforeCritical:         intVal(v.FailuresBeforeCritical),
+		FailuresBeforeWarning:          intValWithDefault(v.FailuresBeforeWarning, intVal(v.FailuresBeforeCritical)),
```

Comment:
This default value looks correct for checks from the config, but unfortunately we have to handle it in different places for the API.

In `HTTPHandlers.AgentRegisterService` and `HTTPHandlers.AgentRegisterCheck` I think we'll need to have a line that sets those to a default value.

Alternatively, maybe we could set that default later on in `NewStatusHandler` ? I honestly don't have a good sense of the trade-offs for those two options. 

As you noticed, since a value of `1` is essentially the same as `0` (if I understand it correctly), maybe we could treat `0` as unset?

Reply:
I don't know if I like using things like special/magic values, especially if they look "correct", like `0`.

I think it's better just to handle individual cases wherever they appear, to be more explicit closes to where they are read.

Reply:
I can see that we are already validating checks inside of `HTTPHandlers.AgentRegisterService`:
https://github.com/hashicorp/consul/blob/710bd90ef76b1d1cf9178f58a7db7b42808021ae/agent/agent_endpoint.go#L990-L996
And we do it inside of `HTTPHandlers.AgentRegisterCheck` as well:
https://github.com/hashicorp/consul/blob/710bd90ef76b1d1cf9178f58a7db7b42808021ae/agent/agent_endpoint.go#L621-L628
So at least validation is covered since `CheckTypes()` does call `Validate()`:
https://github.com/hashicorp/consul/blob/710bd90ef76b1d1cf9178f58a7db7b42808021ae/agent/structs/service_definition.go#L127-L142

Reply:
Truth be told I'm testing setting just `FailuresBeforeCritical` to a value of `6` and it doesn't seem to work at all.
My test service goes into `critical` state essentially right after creation:
```
2021-08-12T10:42:12.185Z [INFO]  agent: Synced check: check=conusl-check-api-call-test
2021-08-12T10:42:23.940Z [WARN]  agent: Check socket connection failed: check=conusl-check-api-call-test error="dial tcp 127.0.0.1:1234: connect: connection refused"
2021-08-12T10:42:23.940Z [WARN]  agent: Check is now critical: check=conusl-check-api-call-test
```
I'm not entirely sure this functionality works at all... it seems like `FailuresBeforeCritical` is being ignored.

Reply:
It doesn't seem to return `FailuresBeforeCritical` from the `GET` query at all:
```
 > curl -s http://localhost:8500/v1/agent/checks | jq '."conusl-check-api-call-test"' 
{
  "Node": "node-01.do-ams3.eth.test",
  "CheckID": "conusl-check-api-call-test",
  "Name": "conusl-check-api-call-test",
  "Status": "critical",
  "Notes": "",
  "Output": "dial tcp 127.0.0.1:1234: connect: connection refused",
  "ServiceID": "",
  "ServiceName": "",
  "ServiceTags": [],
  "Type": "tcp",
  "Interval": "30s",
  "Timeout": "2s",
  "ExposedPort": 0,
  "Definition": {},
  "CreateIndex": 0,
  "ModifyIndex": 0
}
```

Reply:
I've added some debugging prints inside of `StatusHandler.updateCheck` in `agent/checks/check.go`:
```go
		fmt.Printf("checkID.String(): %s\n", checkID.String())
		fmt.Printf("s.failuresCounter: %d\n", s.failuresCounter)
		fmt.Printf("s.failuresBeforeWarning: %d\n", s.failuresBeforeWarning)
		fmt.Printf("s.failuresBeforeCritical: %d\n", s.failuresBeforeCritical)
```
And this is what I got in logs:
```
2021-08-12T11:07:50.077Z [WARN]  agent: Check socket connection failed: check=conusl-check-api-call-test error="dial tcp 127.0.0.1:1234: connect: connection refused"
checkID.String(): conusl-check-api-call-test
s.failuresCounter: 6
s.failuresBeforeWarning: 0
s.failuresBeforeCritical: 6
2021-08-12T11:07:50.077Z [WARN]  agent: Check is now critical: check=conusl-check-api-call-test
```
It appears like `failuresCounter` starts off with value of `6`, which causes it to go into `critical` right away.
https://github.com/hashicorp/consul/blob/710bd90ef76b1d1cf9178f58a7db7b42808021ae/agent/checks/check.go#L921-L922
Not sure if this is the intended behavior but okay.

Reply:
Okay, I see the comment, makes sense:
```go
// NewStatusHandler set counters values to threshold in order to immediatly update status after first check.
```

Reply:
Because `AgentRegisterCheck` loads things into `structs.CheckDefinition` then we can't do what the config does:
https://github.com/hashicorp/consul/blob/8f5e2a440b3a858b224a519ba3b2860386f76f14/agent/structs/check_definition.go#L43-L44
Since the values are of type `int` and not `*int` we won't get `nil` for undefined values. So the only solution is indeed to threat the value of `0` as "undefined", and adjust the default based on that.

Reply:
Also, I noticed that the `OutputMaxSize` attribute has a default:
https://github.com/hashicorp/consul/blob/8f5e2a440b3a858b224a519ba3b2860386f76f14/agent/config/builder.go#L1585
Which is described in the API docs:
![image](https://user-images.githubusercontent.com/2212681/129226915-3f9b089c-3ef8-4520-8013-4ce12f9ec11e.png)
But I do not see it being changed inside of `agent/agent_endpoint.go`. I will follow that trail then, must happen somewhere.

Reply:
It also quite possible that we don't properly default that field. It is known that because there are multiple paths we have a few bugs where different paths behave differently.

Reply:
I see, this is being adjusted inside of `Agent.addCheck`:
https://github.com/hashicorp/consul/blob/525bc2d76366dc52fa464f72a406109447c2ccce/agent/agent.go#L2436
https://github.com/hashicorp/consul/blob/525bc2d76366dc52fa464f72a406109447c2ccce/agent/agent.go#L2465-L2471
So that's probably the right place to add this default.

Reply:
There we go:
https://github.com/status-im/consul/blob/1df254c928e7a0135089ecfabb460fa3a110effe/agent/agent.go#L2468-L2471